### PR TITLE
dcp-944 optimise memory consumption - shallow hashCode

### PR DESCRIPTION
**Pull request by amnonkhen**
change `hashCode` not to traverse DBRef fields to improve efficiency of hashcode calculation by preventing unnecessary db trips.

dcp-944
ebi-ait/dcp-ingest-central#944

### CODE REVIEW

File: src/main/java/org/humancellatlas/ingest/process/Process.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
  * Created by rolando on 16/02/2018.
  */
 @Getter
-@EqualsAndHashCode(callSuper = true)
+@EqualsAndHashCode(callSuper = true, exclude = {"project", "projects", "protocols", "inputBundleManifests", "chainedProcesses"})
```
Comment (aaclan-ebi) [inline src/main/java/org/humancellatlas/ingest/process/Process.java:27]: I don't think the chainedProcesses are being used at all, could just remove that field in the future.

**Review by aaclan-ebi — APPROVED** [review]

**Review by raskoleinonen — COMMENTED** [review]
Looks good to me. All the @DBRef fields have been excluded from @EqualsAndHashCode
of which all were @DBRef(lazy = true) except chainedProcesses.

**Review by raskoleinonen — COMMENTED** [review]
Looks good to me. All the @DBRef fields have been excluded from @EqualsAndHashCode
of which all were @DBRef(lazy = true) except chainedProcesses.

**Comment by tburdett** [comment]
I'm late to the party (so no need to roll back the merge), but out of interest, is there a good reason why the Lombok `@EqualsAndHashCode` annotation is being used at all? `callSuper = true` forces the autogeneration of methods up the inheritence tree, but when I chased it up to `AbstractEntity` I couldn't find any value being added. 

The Lombok docs say this about `@EqualsAndHashCode`: 

> "By default, it'll use all non-static, non-transient fields". 

It seems to me that this default behaviour is the problem (as it uses a number of fields associated with dbrefs, triggering a lazy fetch). You have now fixed this for Processes but it's possible the same issue may pop up elsewhere. 

The `AbstractEntity` class exists to capture common elements for entities in ingest and includes an `Id` field: this seems like an ideal candidate to base all equals and hashcode implementations around! Why not resolve this issue permanently using this suggestion in the Lombok docs and including ONLY the `id` field on `AbstractEntity`? 

> you can specify exactly which fields or methods you wish to be used by marking them with @EqualsAndHashCode.Include and using @EqualsAndHashCode(onlyExplicitlyIncluded = true).

**Comment by amnonkhen** [comment]
Thanks @tburdett @raskoleinonen @aaclan-ebi .
This is a very useful discussion. 
Regarding the suggestion to define the `hashCode` in the super class, that is a cleaner solution. I will, however, test the efficacy and correctness of my initial fix.

**Pull request merged**